### ISSUE TITLE
ANI values not symmetric?

### ISSUE THREAD

**Issue by donovan-h-parks** [bluenote-1577/skani#16]
Hi. I was under the impression that the ANI values produced by skani should be symmetric. This is generally the case, but we have found a few counter examples. The most extreme of these is below:

```
> skani dist -q GCF_009730655.1_ASM973065v1_genomic.fna.gz -r GCF_018799045.1_ASM1879904v1_genomic.fna.gz
[00:00:00.000] (7f195de73800) INFO   skani dist -q GCF_009730655.1_ASM973065v1_genomic.fna.gz -r GCF_018799045.1_ASM1879904v1_genomic.fna.gz
[00:00:00.107] (7f195de73800) INFO   Generating sketch time: 0.10736146
[00:00:00.120] (7f195de73800) INFO   Learned ANI mode detected. ANI may be adjusted according to a regression model trained on MAGs.
Ref_file	Query_file	ANI	Align_fraction_ref	Align_fraction_query	Ref_name	Query_name
GCF_018799045.1_ASM1879904v1_genomic.fna.gz	GCF_009730655.1_ASM973065v1_genomic.fna.gz	93.86	79.68	79.68	NZ_CP076373.1 Shewanella indica strain Colony474 chromosome	NZ_CP046378.1 Shewanella algae strain RQs-106 chromosome, complete genome
[00:00:00.120] (7f195de73800) INFO   ANI calculation time: 0.012946309

> skani dist -r GCF_009730655.1_ASM973065v1_genomic.fna.gz -q GCF_018799045.1_ASM1879904v1_genomic.fna.gz
[00:00:00.000] (7fd2d979a800) INFO   skani dist -r GCF_009730655.1_ASM973065v1_genomic.fna.gz -q GCF_018799045.1_ASM1879904v1_genomic.fna.gz
[00:00:00.106] (7fd2d979a800) INFO   Generating sketch time: 0.106732115
[00:00:00.119] (7fd2d979a800) INFO   Learned ANI mode detected. ANI may be adjusted according to a regression model trained on MAGs.
Ref_file	Query_file	ANI	Align_fraction_ref	Align_fraction_query	Ref_name	Query_name
GCF_009730655.1_ASM973065v1_genomic.fna.gz	GCF_018799045.1_ASM1879904v1_genomic.fna.gz	94.93	79.70	79.70	NZ_CP046378.1 Shewanella algae strain RQs-106 chromosome, complete genome	NZ_CP076373.1 Shewanella indica strain Colony474 chromosome
[00:00:00.119] (7fd2d979a800) INFO   ANI calculation time: 0.012245825
```

Am I mistaken that the ANI values should be symmetric or is this a weird corner case / bug that breaks this assumption?

Thanks,
Donovan

**Comment by bluenote-1577** [comment]
Hi Donovan,

Thanks for calling my attention to this.

I have investigated the files and here are my findings:

1. You're right, skani should be symmetric. The reason it's not here is a bug: when two assemblies have **exactly the same** contig lengths, then skani is no longer symmetric. In this case, these two files have exactly the same contig lengths. Usually, skani smartly picks the query and reference based on contig lengths. 

The assemblers probably used a reference-based scaffolding technique to get these two assemblies to the exact same length...

2. The ANIs are highly non-symmetric. I looked into it, and this is because there **are over 1 million Ns in GCF_018799045.1_ASM1879904v1_genomic** (hopefully I counted correctly). This becomes a thorny situation; how do we treat Ns when it comes to ANI? skani by default turns Ns into As -- probably not the best solution.  

**Technical:** NNNNNN....s get treated differently depending on the if the genome with Ns is the query or the reference. This comes down to the specifics of how skani does k-mer chaining... and if other people are interested, I can talk more about this. 

Anyways, this is why there is an ANI gap when one is the query and the other is the reference. 

**As for solutions:**

Point 1 is easy, this can be remedied easily by checking for equality. 

Point 2 is more tricky, and is also non-trivial to deal with algorithmically. To be honest, this seems more like a data-quality issue, but I think probably the best way to do ANI with Ns is to remove them completely, and shatter the genome into contigs across the Ns. I'll think more about how skani should deal with Ns, but I don't have any final thoughts right now. 

Thanks,

Jim

**Comment by donovan-h-parks** [comment]
Hi Jim,

Thanks for the quick and thorough response. The issue makes sense. Agreed that dealing with N's is a bit tricky, and these genomes clearly aren't the best quality. We only noticed this case since we are processing LOTs of genomes of varying quality. This is a typical use case for us though with the GTDB. Ultimately, the situation is rare enough that it isn't critical, but if the situation can be resolved to ensure skani is always symmetric (and as accurate as possible) that would of course be very welcomed.

Thanks,
Donovan

**Comment by bluenote-1577** [comment]
Hi Donovan,

Thanks for the feedback; what you're saying makes a lot of sense. Glad to hear that skani is seeing usage with the GTDB!

I'll do some testing with genomes with Ns in the near future and decide how to best handle them. I'll keep this issue up until I fix the Ns and symmetry issue, probably for the next release of skani, and I'll update you then.

Thanks,

Jim

**Comment by bluenote-1577** [comment]
@donovan-h-parks I've pushed a new version to this repository that fixes the symmetry issue. If you want to use it, you will have to build from source; see the README. I will release it as v0.1.5 on conda when I've done more testing. 

It turns out that the NNNs issue is surprisingly thorny. For your supplied genomes, if I took the GCF_018799045.1_ASM1879904v1_genomic genome and turned the scaffold into a set of contigs by removing Ns (https://www.biostars.org/p/211400/), MUMmer's dnadiff identity is ~94.8. When I leave it in original, scaffolded form, MUMmer gives ~94.2 ANI. The ~94.9 ANI for skani matches the interpretation after removing Ns and fragmenting into contigs. The new version of skani will give 94.9 ANI here. 

I don't know what MUMmer is doing under the hood here, but wanted to point out that even alignment based ANI methods, which should be robust, deal with Ns poorly. 

Feel free to test out the new version and give feedback.  I'll release this new version in conda the near future.
